mtt/src/mtt.rs

```diff
@@ -0,0 +1,1190 @@
+#[cfg(test)]
+mod mtt_const {
+    pub const L2_TYPE_SHIFT: u64 = 38;
+    // Bits 63:40 must be zero
+    pub const L2_ZERO_MASK: u64 = 0xffff_ff00_0000_0000;
+    // Bits 39:38
+    pub const L2_TYPE_MASK: u64 = 0x00c0_0000_0000;
+    // Bits 37:0
+    pub const L2_INFO_MASK: u64 = 0x003f_ffff_ffff;
+    // Bits 39:38 for a 1G non-confidential region
+    // By convention, this is a total of 16 identical consecutive entries, each mapping 64MB
+    pub const L2_1G_NC_TYPE: u64 = 0;
+    // Bits 39:38 for a 1G confidential region
+    // By convention, this is a total of 16 identical consecutive entries, each mapping 64MB
+    pub const L2_1G_C_TYPE: u64 = 1;
+    // Bits 39:38 for a 64-MB region mapped using a bit-mask in a 4K L1-page
+    pub const L2_L1_TYPE: u64 = 2;
+    // Bits 39:38 for a 64-MB region mapped using a 2MB sub-region bitmask
+    pub const L2_64M_TYPE: u64 = 3;
+    // L2 has 1MB entries
+    pub const MTT_L2_ENTRY_COUNT: usize = 1024 * 1024;
+    // L2 is 8MB long
+    pub const L2_BYTE_SIZE: usize = 1024 * 1024 * 8;
+    // A L1 page consists of 512 8-byte entries
+    pub const L1_ENTRY_COUNT: usize = 512;
+    // Bits 25:17 for the 9-bit index
+    pub const L1_INDEX_MASK: u64 = 0x03fe_0000;
+    pub const L1_INDEX_SHIFT: u64 = 17;
+    // Bits 16:12 for the 5-bit sub-index
+    pub const L1_SUBINDEX_MASK: u64 = 0x1_f000;
+    // 1-bit left shift for the 2-bits per entry for bits (16:12)
+    pub const L1_SUBINDEX_SHIFT: u64 = 11;
+    // 2-bits per entry
+    pub const L1_ENTRY_MASK: u64 = 0x3;
+}
+
+use mtt_const::*;
+
+// Struct to represent a L1 page
+// Each 8-byte entry represents 32 4K pages, and the 512 total entries span a total of
+// 64MB of address space.Each page uses a 2-bit encoding, with '00b as non-confidential
+// and '01b as confidential ('11b is illegal).
+struct MttL1Page<'a> {
+    base: [u64; L1_ENTRY_COUNT],
+    marker: PhantomData<&'a u64>,
+}
+
+impl<'a> MttL1Page<'a> {
+    // Returns a u64 slice that allows access L1 page entries represented by pfn.
+    // #Safety:
+    // The caller must guarantee that the pfn points to a valid L1 page that isn't
+    // aliased elsewhere, and points to confidential memory. The page must remain
+    // allocated for the lifetime of the entry in the L2 table.
+    unsafe fn l1_slice_from_pfn(pfn: SupervisorPfn) -> &'a mut [u64] {
+        let l1_page_addr = (pfn.bits() << 12) as *mut MttL1Page;
+        let l1_page = &mut *core::ptr::slice_from_raw_parts_mut(l1_page_addr, 1);
+        &mut l1_page[0].base[0..]
+    }
+}
+
+/// Holds references to the platform allocated L2 and L2 tables in a Mutex.
+/// This should be constructed by calling init() with the platform allocated backing
+/// memory for the L2 and L1 tables.
+pub struct Mtt {
+    inner: Mutex<MttL2Directory>,
+}
+
+// Singleton instance of Mtt constructed by Mtt::init()
+static PLATFORM_MTT_TABLES: Once<Mtt> = Once::new();
```

Comment:
does the `mtt` module really need to hold the global singleton? couldn't we put it in whatever piece of code (something near `main`) that actually uses it? in other words, this library just handles creating a manipulating an MTT created from a set of pages representing an L2 directory.

the reason i suggest this is that writing tests against a library that uses a singleton like this is painful. you only have one test currently, but once you add more things will quickly blow up due to the tests running in a single process and being multi-threaded within a module.

Reply:
(sorry, should've mentioned this one before too. i recently tried to convert `PageTracker` to a singleton like this and it ended up being more trouble than it was worth due to the testing issue i mentioned)

Reply:
As I see it the issue is that we need to reference the singleton in get(), and moving it upstream will make that difficult?